### ISSUE TITLE
Session's initializer with custom URLSession and SessionDelegate

### ISSUE THREAD

**Issue by rpassis** [Alamofire/Alamofire#3089]
* [X] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What did you do?

When initializing a non-default instance of `Session`, we can provide custom instances of `URLSession` and `SessionDelegate`. 

As far as I can tell `SessionDelegate` will only work as intended if it is set as the delegate for `URLSession`. 

The problem is that the initializer of `Session` does not enforce this requirement so I can pass a `URLSession` instance that is configured with a delegate that is different to the `SessionDelegate` instance passed in the same initializer. In which case `SessionDelegate` would never receive any of the delegate callbacks from `URLSession`.

The convenience initializer does not suffer from the same problem since it creates the `URLSession` instance internally, and configures the `SessionDelegate` passed in as its delegate.

Here's an example:

```swift
let delegateA = SessionDelegate()
let delegateB = SessionDelegate()
let urlSession = URLSession(configuration: configuration, delegate: delegateA, ...)
let afSession = Session(urlSession: urlSession, sessionDelegate: delegateB, ...)
// Delegate B will never receive any URLSession callbacks
```

## What did you expect to happen?

Perhaps we should add a `precondition` for `urlSession.delegate === sessionDelegate`

## What happened instead?

It is possible for both delegates to co-exist, resulting in unintended/unexpected behavior.

## Alamofire Environment

**Alamofire version:** 5.0.0
**Xcode version:** 11.3.1
**Swift version:** 5.1
**Platform(s) running Alamofire:** iOS/macOS
**macOS version running Xcode:** 10.15.3

**Comment by jshier** [comment]
Your suggested `precondition` was originally implemented but was removed in PR #2783 due to an inherent conflict with the swizzling performed by various analytics libraries. The unfortunate popularity of such libraries means that Alamofire can't guarantee the identity of the delegate attached to the `URLSession`, despite whatever `SessionDelegate` value is passed. Various workarounds to allow the `precondition` by default but which would allow users to disable it, perhaps through a build flag, were discussed but rejected due to complexity or inelegance. In the end, due to the nature of the workarounds and the anticipated rarity of initializing with an existing `URLSession`,  we decided to wait and see how common the scenario you describe actually is once users start adopting Alamofire 5 before reconsidering. 

Thanks for looking out! Cheers! 🍻
